crypto/rsa_extra/rsassa_pss_asn1.c

```diff
@@ -0,0 +1,317 @@
+    *mga = RSA_MGA_IDENTIFIER_new();
+    if (*mga != NULL) {
+      (*mga)->mask_gen = mgf1;
+      (*mga)->one_way_hash = hash_algor;
+      return 1;
+    }
+  }
+  OPENSSL_PUT_ERROR(RSA, EVP_R_DECODE_ERROR);
+  RSA_ALGOR_IDENTIFIER_free(mgf1);
+  RSA_ALGOR_IDENTIFIER_free(hash_algor);
+  return 0;
+}
+
+// get_context_specific_value sets |*out| to the contents of DER-encoded, ASN.1
+// element (not including tag and length bytes) and advances |seq| over it.
+// When the tag value does not exist, |seq| gets recovered.
+// It returns one when the element exists.
+static int get_context_specific_value(CBS *seq, CBS *out, int index) {
+  unsigned int tag_value = CBS_ASN1_CONTEXT_SPECIFIC | CBS_ASN1_CONSTRUCTED | index;
+  CBS seq_cp = {seq->data, seq->len};
+  if (CBS_get_asn1(seq, out, tag_value)) {
+    return 1;
+  } else {
+    // All fields of RSASSA-PSS-params can be absent, which means default.
+    // Recover |seq|.
+    seq->data = seq_cp.data;
+    seq->len = seq_cp.len;
+    return 0;
+  }
+}
+
+// Decode [0] HashAlgorithm of RSASSA-PSS-params: return one on success and zero
+// on failure. See 3.1. https://tools.ietf.org/html/rfc4055#page-7
+static int decode_pss_hash(CBS *seq, RSA_ALGOR_IDENTIFIER **hash_algor) {
+  CBS cs;
+  if (!get_context_specific_value(seq, &cs, 0)) {
+    // HashAlgorithm field can be absent, which means default.
+    return 1;
+  }
+  return decode_one_way_hash(&cs, hash_algor);
+}
+
+// Decode [1] MaskGenAlgorithm of RSASSA-PSS-params: return one on success and
+// zero on failure. See 3.1. https://tools.ietf.org/html/rfc4055#page-7
+static int decode_pss_mask_gen(CBS *seq, RSA_MGA_IDENTIFIER **mga) {
+  CBS cs;
+  if (!get_context_specific_value(seq, &cs, 1)) {
+    // MaskGenAlgorithm field can be absent, which means default.
+    return 1;
+  }
+  return decode_mask_gen(&cs, mga);
+}
+
+static int parse_rsa_int(CBS *cbs, RSA_INTEGER **rsa_int) {
+  int64_t value = 0;
+  if (CBS_get_asn1_int64(cbs, &value) && CBS_len(cbs) == 0) {
+    *rsa_int = RSA_INTEGER_new();
+    if ((*rsa_int) != NULL) {
+      (*rsa_int)->value = value;
+      return 1;
+    }
+  }
+  OPENSSL_PUT_ERROR(RSA, EVP_R_DECODE_ERROR);
+  return 0;
+}
+
+// Decode [2] saltLength of RSASSA-PSS-params
+// See 3.1. https://tools.ietf.org/html/rfc4055#page-7
+static int decode_pss_salt_len(CBS *seq, RSA_INTEGER **salt_len) {
+  CBS cs;
+  if (!get_context_specific_value(seq, &cs, 2)) {
+    // saltLength field can be absent, which means default.
+    return 1;
+  }
+  return parse_rsa_int(&cs, salt_len);
+}
+
+// Decode [3] trailerField of RSASSA-PSS-params
+// See 3.1. https://tools.ietf.org/html/rfc4055#page-7
+static int decode_pss_trailer_field(CBS *seq, RSA_INTEGER **trailer_field) {
+  CBS cs;
+  if (!get_context_specific_value(seq, &cs, 3)) {
+    // Trailer field can be absent, which means default.
+    return 1;
+  }
+  return parse_rsa_int(&cs, trailer_field);
```

Comment:
The `trailerField` value MUST be 1. Should check that.

https://tools.ietf.org/html/rfc4055#section-3.1
under `trailerField`.

Reply:
This requirement is not testable in this PR because PSS requires encode MUST omit this field, but decode has to accept non absent case. Both Openssl and JDK11 omit this field. I don't want to forge data here.
```
// Trailer field should not be set because it's for encoding only.
  // See https://tools.ietf.org/html/rfc4055#page-8
  // Sec 3.1. -- MUST omit the trailerField field.
```

However, this is still validated in `crypto/rsa_extra/rsassa_pss_asn1.c` https://github.com/awslabs/aws-lc/pull/90/files 
```
int RSASSA_PSS_PARAMS_get(const RSASSA_PSS_PARAMS *pss, const EVP_MD **md,
                      const EVP_MD **mgf1md, int *saltlen) {
...
if (pss->trailer_field && pss->trailer_field->value != 1) {
...
}

Corresponding test input is InvalidPssGetTestInput
```

Reply:
We agree that in the case of rsassa-pss, if the trailer field is present, then it must be 1?
Then, we can also verify that it is 1, if we have parsed it.
Correct me if I am wrong.

Reply:
The precondition is `Implementations that perform signature validation`. During the validation, `RSASSA_PSS_PARAMS_get` will be called.

```
Implementations that
         perform signature generation MUST omit the trailerField field,
         indicating that the default trailer field value was used.
         Implementations that perform signature validation MUST
         recognize both a present trailerField field with value 1 and an
         absent trailerField field.
```

Reply:
In `decode_pss_trailer_field` you know you are parsing a PSS-params structure. So, if the trailer field is not absent, you also know that it MUST encode 1. Whether the field must be absent or not is decide higher up the stack. That is not something, this part of the code knows about.

Reply:
The RFC does not require trailer field decode and validation MUST happen together. The trailer field has been validated in PR#90.

> if the trailer field is not absent, you also know that it MUST encode 1

`trailer_field` type is `struct RSA_int`, which simply wraps `int`. When it's absent, the value is NULL. When it's 1 or other integer, the value will be store in `int`.